### ISSUE TITLE
virtio-GPU native contexts

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#8552]
[How to file a helpful issue](https://www.qubes-os.org/doc/issue-tracking/)

### The problem you're addressing (if any)

All rendering happens in software.  This is fast enough for many uses, but some users need better performance, even at the expense of increased attack surface.

### The solution you'd like

Opt-in GPU acceleration via virtio-GPU native contexts.

### The value to a user, and who that user might be

Users for whom software rendering has unacceptable performance will be able to use Qubes OS, without resorting to PCI pass-through.

**Comment by rwiesbach** [comment]
https://github.com/orgs/QubesOS/projects/17 bundles efforts regarding GPU acceleration in Qubes OS

**Comment by DemiMarie** [comment]
Removing my assignment as this is blocked on third-party Xen patches being upstreamed.

**Comment by WesleyAC** [comment]
@DemiMarie, could you link to those upstream patches?

**Comment by DemiMarie** [comment]
@WesleyAC I don’t think they are public yet.

**Comment by ILOVEPIE** [comment]
Any way I can assist with development?

**Comment by DemiMarie** [comment]
@ILOVEPIE I think the best ways would be:

- Test the Wayland protocol proxy that I am working on (https://github.com/DemiMarie/wayland-proxy-virtwl).
- Add validation code for Wayland protocol violations.
- Create an automated testing & fuzzing framework for the proxy, so that problems (such as failing to report protocol errors) are detected.

**Comment by tarsa** [comment]
> Wait for Intel virtio-GPU native contexts to be merged upstream.

i think it was added in [Add virtio-intel native context](https://gitlab.freedesktop.org/virgl/virglrenderer/-/merge_requests/1384) and [Support virtio-intel native context](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/29870) (other related activity in [virgl/virglrenderer native context search](https://gitlab.freedesktop.org/virgl/virglrenderer/-/merge_requests/?sort=updated_desc&state=all&search=%22native%20context%22&first_page_size=20) and [mesa/mesa native context search](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/?sort=updated_desc&state=all&search=%22native%20context%22&first_page_size=20)).

since qubes os certified laptops have certification for intel graphics and qubes os developers have qubes os certified laptops for ongoing testing, should the recently merged intel virtio-gpu native context support unlock development of gpu acceleration in qubes os (i.e. the mechanisms mentioned in this meta-issue)?

**Comment by DemiMarie** [comment]
The biggest problem right now is that native contexts are not supported with Xen due to hypervisor and kernel limitations.  Fundamentally, Linux kernel memory management requires that GPU memory be able to be moved without userspace noticing, and Xen does not allow guest memory to be paged in and out transparently to the guest.  KVM does, which is why all of this works just fine on KVM.

AMD has managed to make this work with what I am almost certain is a non-upstreamable hack.

**Comment by lyphyser** [comment]
Switch the hypervisor from Xen to KVM?

There used to be issues with the qemu attack surface, but nowadays there is plenty of KVM userspace that is written in Rust so there is no need to use qemu.

**Comment by andrewdavidwong** [comment]
> Switch the hypervisor from Xen to KVM?

There's a separate issue about that: https://github.com/QubesOS/qubes-issues/issues/7051

**Comment by tarsa** [comment]
if xen developers are unwilling to add support for virtio-gpu native contexts, then maybe qubes os could at least support sr-iov on intel igpu? how feasible would that be?

edit:
qemu added support for feature from this issue:
https://wiki.qemu.org/ChangeLog/11.0
> The virtio-gpu now has support for native context drivers

**Comment by Atrate** [comment]
> then maybe qubes os could at least support sr-iov on intel igpu? how feasible would that be?

Some people have had success with SR-IOV on the forums: https://forum.qubes-os.org/t/virtualised-intel-gpu-with-sr-iov/40649
